File: app/src/main/java/com/intelligentquestionbank/memory/fragments/ImportFragment.java
package com.intelligentquestionbank.memory.fragments;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import com.intelligentquestionbank.memory.R;
import com.intelligentquestionbank.memory.database.entities.Question;
import com.intelligentquestionbank.memory.database.entities.QuestionLibrary;
import com.intelligentquestionbank.memory.utils.QuestionParser;
import com.intelligentquestionbank.memory.viewmodels.ImportViewModel;
import java.util.List;

public class ImportFragment extends Fragment {

    private EditText editTextContent;
    private EditText editTextLibraryName;
    private Spinner spinnerExistingLibrary;
    private Button buttonCreateNew;
    private Button buttonUseExisting;
    private Button buttonImport;
    private TextView textViewStats;
    private TextView textViewCharCount;

    private ImportViewModel viewModel;
    private boolean useNewLibrary = true;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, 
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_import, container, false);
        
        initViews(view);
        initViewModel();
        setupListeners();
        observeData();
        
        return view;
    }

    private void initViews(View view) {
        editTextContent = view.findViewById(R.id.edit_text_content);
        editTextLibraryName = view.findViewById(R.id.edit_text_library_name);
        spinnerExistingLibrary = view.findViewById(R.id.spinner_existing_library);
        buttonCreateNew = view.findViewById(R.id.button_create_new);
        buttonUseExisting = view.findViewById(R.id.button_use_existing);
        buttonImport = view.findViewById(R.id.button_import);
        textViewStats = view.findViewById(R.id.text_view_stats);
        textViewCharCount = view.findViewById(R.id.text_view_char_count);
    }

    private void initViewModel() {
        viewModel = new ViewModelProvider(this).get(ImportViewModel.class);
    }

    private void setupListeners() {
        buttonCreateNew.setOnClickListener(v -> {
            useNewLibrary = true;
            updateLibrarySelectionUI();
        });

        buttonUseExisting.setOnClickListener(v -> {
            useNewLibrary = false;
            updateLibrarySelectionUI();
        });

        buttonImport.setOnClickListener(v -> performImport());

        editTextContent.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                updateCharCount(s.length());
                parseAndShowPreview(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });
    }

    private void observeData() {
        viewModel.getAllLibraries().observe(getViewLifecycleOwner(), libraries -> {
            ArrayAdapter<QuestionLibrary> adapter = new ArrayAdapter<>(
                    getContext(), android.R.layout.simple_spinner_item, libraries);
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            spinnerExistingLibrary.setAdapter(adapter);
        });
    }

    private void updateLibrarySelectionUI() {
        if (useNewLibrary) {
            buttonCreateNew.setBackgroundResource(R.drawable.button_primary);
            buttonUseExisting.setBackgroundResource(R.drawable.button_outline);
            editTextLibraryName.setVisibility(View.VISIBLE);
            spinnerExistingLibrary.setVisibility(View.GONE);
        } else {
            buttonCreateNew.setBackgroundResource(R.drawable.button_outline);
            buttonUseExisting.setBackgroundResource(R.drawable.button_primary);
            editTextLibraryName.setVisibility(View.GONE);
            spinnerExistingLibrary.setVisibility(View.VISIBLE);
        }
    }

    private void updateCharCount(int count) {
        textViewCharCount.setText(String.format("字符数: %d / 100000", count));
        textViewCharCount.setTextColor(count > 100000 ? 
                getResources().getColor(R.color.error_color) : 
                getResources().getColor(R.color.secondary_text_color));
    }

    private void parseAndShowPreview(String content) {
        if (content.trim().isEmpty()) {
            textViewStats.setText("预览统计：\n无内容");
            return;
        }

        try {
            List<Question> questions = QuestionParser.parseQuestionsFromText(content);
            int fillBlankCount = 0;
            int shortAnswerCount = 0;

            for (Question question : questions) {
                if ("fill_in_blank".equals(question.getType())) {
                    fillBlankCount++;
                } else {
                    shortAnswerCount++;
                }
            }

            textViewStats.setText(String.format(
                    "预览统计：\n总题数：%d\n填空题：%d\n简答题：%d",
                    questions.size(), fillBlankCount, shortAnswerCount));

        } catch (Exception e) {
            textViewStats.setText("解析错误：" + e.getMessage());
        }
    }

    private void performImport() {
        String content = editTextContent.getText().toString().trim();
        if (content.isEmpty()) {
            Toast.makeText(getContext(), "请输入题目内容", Toast.LENGTH_SHORT).show();
            return;
        }

        if (content.length() > 100000) {
            Toast.makeText(getContext(), "内容超过10万字符限制", Toast.LENGTH_SHORT).show();
            return;
        }

        QuestionLibrary targetLibrary;
        if (useNewLibrary) {
            String libraryName = editTextLibraryName.getText().toString().trim();
            if (libraryName.isEmpty()) {
                Toast.makeText(getContext(), "请输入题库名称", Toast.LENGTH_SHORT).show();
                return;
            }
            targetLibrary = new QuestionLibrary();
            targetLibrary.setName(libraryName);
            targetLibrary.setDescription("通过文本导入创建");
            targetLibrary.setCreatedAt(System.currentTimeMillis());
        } else {
            targetLibrary = (QuestionLibrary) spinnerExistingLibrary.getSelectedItem();
            if (targetLibrary == null) {
                Toast.makeText(getContext(), "请选择目标题库", Toast.LENGTH_SHORT).show();
                return;
            }
        }

        buttonImport.setEnabled(false);
        buttonImport.setText("导入中...");

        viewModel.importQuestions(content, targetLibrary, success -> {
            if (getActivity() != null) {
                getActivity().runOnUiThread(() -> {
                    buttonImport.setEnabled(true);
                    buttonImport.setText("开始导入");
                    
                    if (success) {
                        Toast.makeText(getContext(), "导入成功！", Toast.LENGTH_SHORT).show();
                        editTextContent.setText("");
                        editTextLibraryName.setText("");
                    } else {
                        Toast.makeText(getContext(), "导入失败，请检查格式", Toast.LENGTH_SHORT).show();
                    }
                });
            }
        });
    }
}

File: app/src/main/java/com/intelligentquiz/memory/fragments/ImportFragment.kt
package com.intelligentquiz.memory.fragments

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.intelligentquiz.memory.R
import com.intelligentquiz.memory.databinding.FragmentImportBinding
import com.intelligentquiz.memory.utils.QuestionParser
import com.intelligentquiz.memory.viewmodels.ImportViewModel
import kotlinx.coroutines.launch

class ImportFragment : Fragment() {
    
    private var _binding: FragmentImportBinding? = null
    private val binding get() = _binding!!
    
    private lateinit var viewModel: ImportViewModel
    private var isCreatingNewLibrary = false
    
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentImportBinding.inflate(inflater, container, false)
        return binding.root
    }
    
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        
        viewModel = ViewModelProvider(this)[ImportViewModel::class.java]
        
        setupSpinner()
        setupTextWatcher()
        setupClickListeners()
        observeViewModel()
    }
    
    private fun setupSpinner() {
        // Load existing libraries and add "创建新题库" option
        viewModel.loadLibraries()
        
        viewModel.libraries.observe(viewLifecycleOwner) { libraries ->
            val spinnerItems = mutableListOf<String>()
            spinnerItems.add("创建新题库")
            spinnerItems.addAll(libraries.map { it.name })
            
            val adapter = ArrayAdapter(
                requireContext(),
                android.R.layout.simple_spinner_item,
                spinnerItems
            )
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            binding.spinnerTargetLibrary.adapter = adapter
            
            binding.spinnerTargetLibrary.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                    isCreatingNewLibrary = position == 0
                    binding.editTextNewLibraryName.visibility = if (isCreatingNewLibrary) {
                        View.VISIBLE
                    } else {
                        View.GONE
                    }
                }
                
                override fun onNothingSelected(parent: AdapterView<*>?) {}
            }
        }
    }
    
    private fun setupTextWatcher() {
        binding.editTextContent.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            
            override fun afterTextChanged(s: Editable?) {
                val content = s?.toString() ?: ""
                updateCharCount(content.length)
                updatePreviewStats(content)
            }
        })
    }
    
    private fun setupClickListeners() {
        binding.buttonImport.setOnClickListener {
            performImport()
        }
    }
    
    private fun observeViewModel() {
        viewModel.importResult.observe(viewLifecycleOwner) { result ->
            if (result.isSuccess) {
                Toast.makeText(context, "导入成功！共导入${result.getOrNull()}道题目", Toast.LENGTH_LONG).show()
                binding.editTextContent.setText("")
                binding.editTextNewLibraryName.setText("")
                // Return to home fragment
                parentFragmentManager.popBackStack()
            } else {
                Toast.makeText(context, "导入失败：${result.exceptionOrNull()?.message}", Toast.LENGTH_LONG).show()
            }
        }
        
        viewModel.isLoading.observe(viewLifecycleOwner) { isLoading ->
            binding.buttonImport.isEnabled = !isLoading
            binding.buttonImport.text = if (isLoading) "导入中..." else "开始导入"
        }
    }
    
    private fun updateCharCount(count: Int) {
        // Find the char count view in the layout
        // binding.textViewCharCount.text = "字符数: $count / ${QuestionParser.MAX_IMPORT_LENGTH}"
        // For now, we'll skip the char count display since the view references need to be fixed
    }
    
    private fun updatePreviewStats(content: String) {
        lifecycleScope.launch {
            if (content.isEmpty()) {
                // binding.textViewStats.text = "预览统计：\n无内容"
                return@launch
            }
            
            val parseResult = QuestionParser.parseText(content, 0) // temporary library ID
            // binding.textViewStats.text = """
            //     预览统计：
            //     总题数：${parseResult.totalParsed}
            //     填空题：${parseResult.fillInBlankCount}
            //     简答题：${parseResult.shortAnswerCount}
            //     ${if (parseResult.errors.isNotEmpty()) "\n错误：${parseResult.errors.size}处" else ""}
            // """.trimIndent()
        }
    }
    
    private fun performImport() {
        val content = binding.editTextContent.text.toString().trim()
        
        if (content.isEmpty()) {
            Toast.makeText(context, "请输入题目内容", Toast.LENGTH_SHORT).show()
            return
        }
        
        if (content.length > QuestionParser.MAX_IMPORT_LENGTH) {
            Toast.makeText(context, "内容超过${QuestionParser.MAX_IMPORT_LENGTH}字符限制", Toast.LENGTH_SHORT).show()
            return
        }
        
        val libraryName = if (isCreatingNewLibrary) {
            val newName = binding.editTextNewLibraryName.text.toString().trim()
            if (newName.isEmpty()) {
                Toast.makeText(context, "请输入新题库名称", Toast.LENGTH_SHORT).show()
                return
            }
            newName
        } else {
            val selectedPosition = binding.spinnerTargetLibrary.selectedItemPosition
            if (selectedPosition <= 0) {
                Toast.makeText(context, "请选择目标题库", Toast.LENGTH_SHORT).show()
                return
            }
            binding.spinnerTargetLibrary.selectedItem.toString()
        }
        
        viewModel.importQuestions(content, libraryName, isCreatingNewLibrary)
    }
    
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}